File: leetcode/25.py
class Solution(object):
    def reverseKGroup(self, head, k):
        if head is None or k < 2:   return head
        now = head
        for i in xrange(k - 1):
            now = now.next
            if now == None: return head
        pre, cur = None, head
        for i in xrange(k):
            pre, cur.next, cur = cur, pre, cur.next
        head.next = self.reverseKGroup(cur, k)
        return now

File: leetcode/25.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
       if (!head)  return NULL;
       ListNode *now = head;
       for (int i = 0; i < k; ++i) {
           if (now == NULL) return head;
           now = now -> next;
       }
       ListNode *last = NULL, *t = head;
       for (int i = 0; i < k; ++i) {
           ListNode* tmp = t -> next;
           t -> next = last;
           last = t;
           t = tmp;
       }
       head -> next = reverseKGroup(now, k);
       return last; 
    }
};